inc/azure_macro_utils/macro_utils.h

```diff
@@ -78,73 +78,101 @@ MU_IF(X, "true", "false") => "true"
 #define MU_EAT_EMPTY_ARG(arg_count, arg) MU_IF(MU_ISEMPTY(arg),,arg) MU_IF(MU_ISEMPTY(arg),MU_EXPAND_TO_NOTHING,MU_IFCOMMALOGIC)(MU_DEC(arg_count))
 #define MU_EAT_EMPTY_ARGS(...) MU_FOR_EACH_1_COUNTED(MU_EAT_EMPTY_ARG, __VA_ARGS__)
 
+#define MU_EMPTY()
+#define MACRO_UTILS_DELAY(id) id MU_EMPTY MU_LPAREN )
+
+//
+//
+// MU_DEFINE_ENUM
+//
+//
+
+// This uber enum macro can handle enumeration-constants being constant expressions.
+// Given the following enum:
+//
+//
+// #define TEST_UBER_ENUM_VALUES \
+//     test_uber_a, \
+//     test_uber_b = 5
+// 
+// MU_DEFINE_ENUM(TEST_UBER_ENUM, TEST_UBER_ENUM_VALUES)
+//
+// The ToString expands to something like the below given an enum
+//
+// const char* TEST_UBER_ENUM_ToString(TEST_UBER_ENUM value)
+// {
+//     static int my_test_uber_a;
+//     static int my_test_uber_b = 5;
+//     static int enum_last_value = -1;
+//     enum_last_value = my_test_uber_a ? (my_test_uber_a) : (enum_last_value + 1);
+//     if (enum_last_value == value)
+//     {
+//         return "test_uber_a";
+//     }
+//     enum_last_value = (my_test_uber_b = 5) ? (my_test_uber_b = 5) : (enum_last_value + 1);
+//     if (enum_last_value == value)
+//     {
+//         return "test_uber_a";
+//     }
+//     return "NULL";
+// }
+
 #define MU_DEFINE_ENUMERATION_CONSTANT(x) x,
-/*MU_DEFINE_ENUM_WITHOUT_INVALID goes to header*/
-#define MU_DEFINE_ENUM_WITHOUT_INVALID(enumName, ...) typedef enum MU_C2(enumName, _TAG) { MU_FOR_EACH_1(MU_DEFINE_ENUMERATION_CONSTANT, __VA_ARGS__)} enumName; \
-    extern const char* MU_C2(enumName,Strings)(enumName value); \
-    extern int MU_C2(enumName, _FromString)(const char* enumAsString, enumName* destination);
 
-// this macro is a wrapper on top of MU_DEFINE_ENUM_WITHOUT_INVALID, adding an _INVALID value as the first enum value in the enum
+#define MU_INTERNAL_DECLARE_ENUM_VAR(enumValue) \
```

Comment:
>DECLARE [](start = 20, length = 7)

one would argue that this is a definition ./as a matter of poetry

Reply:
Sure, renamed.

---
In reply to: [328360488](https://github.com/Azure/azure-macro-utils-c/pull/30#discussion_r328360488) [](ancestors = 328360488)